c10/core/Allocator.h

```diff
@@ -127,6 +127,10 @@ struct C10_API Allocator {
     AT_ASSERT(d);
     d(ptr);
   }
+  // Used by allocators that maintain state, i.e. CPURegionAllocator
+  virtual void reset() {
+    return;
+  }
```

Comment:
I'm a bit ambivalent about this. Under the current allocator model, all `Allocator` objects are expected to be static lifetime. That implies that there's no need to actually put the state in the Allocator object itself, and generally this is a good thing because the Allocator interface exists purely to let us dynamic dispatch to the allocate function when we don't know what the allocator is; but interesting allocators have all sorts of extra knobs, and you don't want to add all of those knobs to Allocator.

The key litmus test is, will you ever call `reset()` on a generic `Allocator` which you don't know what it is? In the non-OSS portion of this patch, there is a call to reset(), but in a context where we *know* the region allocator is being used. That to me is a strong indicator we shouldn't have this virtual method on this class.

Reply:
OK, looking at CPURegionAllocator, I see that you do, in fact, want to have multiple instances of the Allocator object. So maybe we should support this pattern, but we will have to be careful about how heavily we've made assumptions about this elsewhere. One particular hazard I can think of right off the bat is that if you have a Tensor allocated by CPURegionAllocator, and it persists beyond the scope of the CPURegionAllocator, it will end up with a dangling Allocator pointer (as well as dangling memory, so maybe this is not the... worst thing in the world? But it is pretty nasty for users who expect tensors to stay valid anywhere.) cc @Yangqing who I remember advocated quite strongly for static lifetime allocators.